xpm: Fix various parser panics

gnoack opened the pull request:
This series of patches adds subtests to test input files against golden PNGs, a fuzzer, and then various changes to fix the detected panics in the parser.

gnoack committed:
Use fuzz testing to catch crashes

https://go.dev/doc/security/fuzz/

gnoack committed:
xpm: detect negative image dimensions

gnoack committed:
xpm: detect when pixel data is missing

gnoack committed:
xpm: detect errors during color parsing

The existing code returned nil colors on errors,
which leads to crashes.

gnoack committed:
xpm: detect too large images

gnoack committed:
xpm: detect if there is too much pixel data

gnoack committed:
Generalize parser test and add another example file

gnoack committed:
xpm: add the fuzz reproducers for these various crashes

gnoack committed:
Require Go 1.18 or higher

Fuzzing support was only introduced in Go 1.18.

Jacalz commented on go.mod:
```diff
@@ -1,8 +1,14 @@
 module github.com/fyne-io/image
 
-go 1.14
+go 1.18
```
The upcoming version of Fyne is targeting 1.19 as baseline. You might as well bump it once more.

gnoack replied:
Makes sense, will add a commit for that which you can squash with 5bdced54

Jacalz commented on xpm/xpm.go:
```diff
@@ -43,18 +49,19 @@ func parseXPM(data io.Reader) (image.Image, error) {
-	// instead of just returning and implicitly leaving err as nil.
 	if len(data) < charSize {
-		return
+		return "", nil, fmt.Errorf("invalid format: missing color specification")
```
You are not using any formatting. Consider using `errors.New()` instead. You might also want to bring it out to a global error variable given that the same error is used in multiple places.

gnoack replied:
I introduced an `ErrInvalidFormat` and return `fmt.Errorf("%w: ...")` in the error cases, so that callers can tell it apart.

Jacalz commented on xpm/xpm.go:
```diff
@@ -111,16 +122,41 @@ func parseDimensions(data string) (w, h, i, j int, err error) {
+		return
+	}
+	if ncolors <= 0 {
+		err = fmt.Errorf("invalid format: ncolors <= 0: missing color palette")
```
Consider using `errors.New()`.

gnoack replied:
Using `fmt.Errorf("%w: ...", ErrInvalidFormat)` (see comment above)

Jacalz commented on xpm/xpm.go:
```diff
@@ -111,16 +122,41 @@ func parseDimensions(data string) (w, h, i, j int, err error) {
 	}
-	j, err = strconv.Atoi(parts[3])
+	if cpp <= 0 {
+		err = fmt.Errorf("invalid format: characters per pixel <= 0")
```
Consider using `errors.New()`.

Jacalz commented on xpm/xpm.go:
```diff
@@ -111,16 +122,41 @@ func parseDimensions(data string) (w, h, i, j int, err error) {
-func parsePixels(row string, charSize int, pixRow int, colors map[string]color.Color, img *image.NRGBA) {
+func parsePixels(row string, charSize int, pixRow int, colors map[string]color.Color, img *image.NRGBA) error {
+	if len(row) < charSize*(img.Stride/4) {
+		return fmt.Errorf("invalid format: missing pixel data")
```
Consider using `errors.New()`.

gnoack replied:
Using `fmt.Errorf("%w: ...", ErrInvalidFormat)` (see comment above)

Jacalz commented on xpm/xpm.go:
```diff
@@ -111,16 +122,41 @@ func parseDimensions(data string) (w, h, i, j int, err error) {
+	}
 	off := pixRow * img.Stride
+	if len(img.Pix) < off+img.Stride {
+		return fmt.Errorf("invalid format: too much pixel data")
```
Consider using `errors.New()`.

gnoack replied:
Using `fmt.Errorf("%w: ...", ErrInvalidFormat)` (see comment above)

Jacalz reviewed (changes requested):
Thanks. A very nice change. Added some notes  inline. Mostly about error handling.

Bluebugs commented on xpm/fuzz_test.go:
```diff
@@ -0,0 +1,32 @@
+// parser.  To run, use:
+//
+//	go test -fuzz=FuzzParsePanic
+func FuzzParsePanic(f *testing.F) {
```
I think it would be ok to run fuzz automatically everyday in a github action. I usually pick 4am as that's when the activity on fyne repository are at their lowest. If you do not mind adding a new action, I think that will be useful to run this fuzzing regularly.

gnoack replied:
Fuzzing runs indefinitely if you let it, and I'm not sure whether Github actions are the best fit there.  (TBH, I'm not terribly familiar with Github actions either.)  I suspect that these considerations would inflate the scope of this change too much.  I'd rather leave it up to you to configure this, if that's OK with you. :)

gnoack committed:
Require Go 1.19 or higher

Fixes https://github.com/fyne-io/image/pull/7#discussion_r1459989328

(Should be a fixup on commit 5bdced54.)

gnoack committed:
xpm: Introduce ErrInvalidFormat for input format errors

We return fmt.Errorf("%w: ...") in the error cases,
so callers can tell the error apart using `errors.Is()` and
`errors.As()`.

gnoack commented:
> Thanks. A very nice change. Added some notes inline. Mostly about error handling.

Thanks for the review. I added some commits on top to address these; you might want to squash the go.mod change onto 5bdced54 before merging.  Please take another look.

andydotxyz reviewed (approved):
Nice addition of safety thanks :)

Jacalz commented:
Thanks for working on this :)
